it/src/main/java/com/google/cloud/teleport/it/dataflow/DefaultDataflowClient.java

```diff
@@ -0,0 +1,399 @@
+import com.google.api.services.dataflow.model.LaunchTemplateParameters;
+import com.google.api.services.dataflow.model.LaunchTemplateResponse;
+import com.google.api.services.dataflow.model.MetricUpdate;
+import com.google.auth.Credentials;
+import com.google.auth.http.HttpCredentialsAdapter;
+import com.google.cloud.teleport.it.logging.LogStrings;
+import com.google.common.base.Strings;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Default class for implementation of {@link DataflowClient} interface. */
+public class DefaultDataflowClient implements DataflowClient {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultDataflowClient.class);
+  private static final Pattern JOB_ID_PATTERN = Pattern.compile("Submitted job: (\\S+)");
+  private static final Pattern CURRENT_METRICS = Pattern.compile(".*Current.*");
+  protected final Dataflow client;
+
+  DefaultDataflowClient(Dataflow client) {
+    this.client = client;
+  }
+
+  private DefaultDataflowClient(DefaultDataflowClient.Builder builder) {
+    this(
+        new Dataflow(
+            Utils.getDefaultTransport(),
+            Utils.getDefaultJsonFactory(),
+            new HttpCredentialsAdapter(builder.getCredentials())));
+  }
+
+  public static DefaultDataflowClient withDataflowClient(Dataflow client) {
+    return new DefaultDataflowClient(client);
+  }
+
+  public static DefaultDataflowClient.Builder builder() {
+    return new DefaultDataflowClient.Builder();
+  }
+
+  @Override
+  public JobInfo launchFlexTemplate(String project, String region, LaunchConfig options)
+      throws IOException {
+    checkState(
+        options.specPath() != null,
+        "Cannot launch a template job without specPath. Please specify specPath and try again!");
+    LOG.info("Getting ready to launch {} in {} under {}", options.jobName(), region, project);
+    LOG.info("Using the spec at {}", options.specPath());
+    LOG.info("Using parameters:\n{}", formatForLogging(options.parameters()));
+
+    LaunchFlexTemplateParameter parameter =
+        new LaunchFlexTemplateParameter()
+            .setJobName(options.jobName())
+            .setParameters(options.parameters())
+            .setContainerSpecGcsPath(options.specPath());
+    LaunchFlexTemplateRequest request =
+        new LaunchFlexTemplateRequest().setLaunchParameter(parameter);
+    LOG.info("Sending request:\n{}", formatForLogging(request));
+
+    LaunchFlexTemplateResponse response =
+        client.projects().locations().flexTemplates().launch(project, region, request).execute();
+    LOG.info("Received response:\n{}", formatForLogging(response));
+
+    Job job = response.getJob();
```

Comment:
We had some logic to print the console URL to the job that was really useful for debugging, consider putting it back if possible

Reply:
Sorry, added that back